### PR TITLE
8330853: Add missing checks for ConnectionGraph::can_reduce_cmp() call

### PR DESCRIPTION
**Pull request by vnkozlov**
In Leyden testing CI we start hitting assert:

```
# Internal Error (/workspace/open/src/hotspot/share/opto/node.hpp:407), pid=3216007, tid=3216030
# assert(i < _max) failed: oob: i=2, _max=2

Stack: [0x00007f20b011b000,0x00007f20b021b000], sp=0x00007f20b02157e0, free space=1001k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V [libjvm.so+0xbf5bb4] Node::in(unsigned int) const [clone .part.0]+0x24 (node.hpp:407)
V [libjvm.so+0xbf69ea] ConnectionGraph::can_reduce_cmp(Node*, Node*) const+0x9a (node.hpp:407)
V [libjvm.so+0xbf74a6] ConnectionGraph::can_reduce_check_users(Node*, unsigned int) const+0x996 (escape.cpp:578)
```

[JDK-8316991](https://bugs.openjdk.org/browse/JDK-8316991) added new code which assumes that we always have sequence : IfNode->Bool->Cmp: [escape.cpp#L569](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/escape.cpp#L569)
which is not true in some cases. In failed cases there is additional Opaque4 node between If and Bool nodes.

The fix is to add missing checks for If, Bool and Cmp nodes.

The fix is verified in Leyden CI testing. The failure happens only with special C2 mode compilation in Leyden.
Running our regular tier1-3,stress,xcomp too.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8330853](https://bugs.openjdk.org/browse/JDK-8330853): Add missing checks for ConnectionGraph::can_reduce_cmp() call (**Bug** - P3)

### Reviewers
 * [Igor Veresov](https://openjdk.org/census#iveresov) (@veresov - **Reviewer**)
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)
 * [Cesar Soares Lucas](https://openjdk.org/census#cslucas) (@JohnTortugo - Author)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18916/head:pull/18916` \
`$ git checkout pull/18916`

Update a local copy of the PR: \
`$ git checkout pull/18916` \
`$ git pull https://git.openjdk.org/jdk.git pull/18916/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18916`

View PR using the GUI difftool: \
`$ git pr show -t 18916`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18916.diff">https://git.openjdk.org/jdk/pull/18916.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18916#issuecomment-2072518023)

### CODE REVIEW

**Comment by vnkozlov** [comment]
linux-x86 failed to upload results but tests passed:
```
2024-04-23T15:24:09.6443002Z    TEST                                    TOTAL  PASS  FAIL ERROR   
2024-04-23T15:24:09.6443822Z    jtreg:test/hotspot/jtreg:tier1_gc         312   312     0     0   
```

**Comment by vnkozlov** [comment]
@JohnTortugo, please look.

File: src/hotspot/share/opto/escape.cpp [hunk 1]
```diff
@@ -568,11 +568,14 @@ bool ConnectionGraph::can_reduce_check_users(Node* n, uint nesting) const {
...
-          if (!can_reduce_cmp(n, iff_cmp)) {
-            NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d on invocation %d. CastPP %d doesn't have simple control.", n->_idx, _invocation, use->_idx);)
-            NOT_PRODUCT(n->dump(5);)
-            return false;
+          if (iff->Opcode() == Op_If && iff->in(1)->is_Bool() && iff->in(1)->in(1)->is_Cmp()) {
+            Node* iff_cmp = iff->in(1)->in(1);
+            int opc = iff_cmp->Opcode();
+            if ((opc == Op_CmpP || opc == Op_CmpN) && !can_reduce_cmp(n, iff_cmp)) {
+              NOT_PRODUCT(if (TraceReduceAllocationMerges) tty->print_cr("Can NOT reduce Phi %d on invocation %d. CastPP %d doesn't have simple control.", n->_idx, _invocation, use->_idx);)
+              NOT_PRODUCT(n->dump(5);)
```
Comment (veresov) [inline src/hotspot/share/opto/escape.cpp:576]: It's a preexisting problem but do you want to call `dump()` unconditionally? Shouldn't it be under `if (TraceReduceAllocationMerges)` ?

**Comment by vnkozlov** [comment]
I found only one other place where `dump()` was not under flag. Fixed.

**Review by veresov — APPROVED** [review]
Looks good

**Comment by vnkozlov** [comment]
Thank you, Igor

**Review by dean-long — APPROVED** [review]

**Comment by vnkozlov** [comment]
Thank you, Dean, for review.

**Comment by vnkozlov** [comment]
GHA failure in linux-x86 (failed compilation of test/langtools/tools/javac/lambda/TargetType69.java test) looks unrelated to my changes.

**Review by JohnTortugo — APPROVED** [review]

**Comment by vnkozlov** [comment]
Thank you, @JohnTortugo

**Pull request closed without merging**